### DISCUSSION TITLE
xworkspaces: nwin counting its own polybar window

### DISCUSSION THREAD

**Discussion by mattmart3** [polybar/polybar#3125] [General]
I am asking here since I am not yet sure it is a bug or not. I am using the `%nwin%` counter with xworkspaces labels (e.g. `label-active = %name%: %nwin%`) and I see that the polybar window itself gets added to the counter.
In the example below I have 3 windows on first desktop: firefox, alacritty and polybar (shown as N/A because it has not WM_NAME apparently) and 1 window on second desktop: telegram.
```
wmctrl -l
0x00c0002d  0 njaxe New Discussion · polybar/polybar — Mozilla Firefox
0x01200003  0 njaxe Zellij (lucky-apricot) - eugene@njaxe:~/dev/polybar
0x01000001  0   N/A
0x01c0000a  1 njaxe Telegram
```
![image](https://github.com/polybar/polybar/assets/1726300/a64c9faf-cfcf-4be7-a99d-14945c43469a)

I expected the polybar window not to be counted. Is this instead the expected behavior? Is there any launch or config option to disable this?

With the following workaround I can get rid of it:

```diff
diff --git a/src/modules/xworkspaces.cpp b/src/modules/xworkspaces.cpp
index 57a6fef2..dbea3c36 100644
--- a/src/modules/xworkspaces.cpp
+++ b/src/modules/xworkspaces.cpp
@@ -153,6 +153,10 @@ namespace modules {
     m_clients.clear();
     m_windows.clear();
     for (auto&& client : newclients) {
+      if (icccm_util::get_wm_class(m_connection, client).second.compare(("Polybar")) == 0) {
+        // Skip Polybar window
+        continue;
+      }
       auto desk = ewmh_util::get_desktop_from_window(client);
       m_clients[client] = desk;
       m_windows[desk]++;
```
![image](https://github.com/polybar/polybar/assets/1726300/3196a873-0a85-48e0-a973-ce5e238a8397)
Notice the counter on first destkop is now 2 instead of 3.

Is there anyone using %nwin% that can confirm this? Tagging the original author @BrunoCooper17.

**Comment by patrick96** [comment]
What WM is this?

For some reason polybar gets attributed to the first workspace even though polybar requests to be visible on all workspaces.
This may be due to your window manager not properly supporting `_NET_WM_DESKTOP`.
Normally, polybar would be attributed to workspace -1, which doesn't exist, so it should not show up on `%nwin%`

I'm assuming it is visible on all workspaces, right?

Can you also run `xprop`, click on your polybar, and post the output?

**Reply by mattmart3** [reply]
> What WM is this?

[LeftWM](https://github.com/leftwm/leftwm). 
Polybar is launched from an `up` script, something similar to [this](https://github.com/JacoMalan1/LateNight/blob/a70d123e74e92d8c9a34b041b5865f4d9f7fa983/up#L30). 

> For some reason polybar gets attributed to the first workspace even though polybar requests to be visible on all workspaces. This may be due to your window manager not properly supporting `_NET_WM_DESKTOP`. Normally, polybar would be attributed to workspace -1, which doesn't exist, so it should not show up on `%nwin%`

It looks like polybar gets attributed to the current workspace if I launch it manually: e.g. if I move to workspace 2 and launch polybar it is added on the nwin counter of workspace 2.

> I'm assuming it is visible on all workspaces, right?

Yes.

> Can you also run `xprop`, click on your polybar, and post the output?

```
$ xprop
WM_STATE(WM_STATE):
                window state: Normal
                icon window: <field not available>
WM_NORMAL_HINTS(WM_SIZE_HINTS):
                program specified location: 0, 0
                program specified size: 2560 by 28
                program specified minimum size: 2560 by 28
                program specified maximum size: 2560 by 28
                program specified base size: 2560 by 28
_NET_WM_PID(CARDINAL) = 41250
_NET_WM_DESKTOP(CARDINAL) = 1
_NET_WM_STATE(ATOM) = _NET_WM_STATE_STICKY, _NET_WM_STATE_ABOVE
_NET_WM_WINDOW_TYPE(ATOM) = _NET_WM_WINDOW_TYPE_DOCK
WM_CLASS(STRING) = "polybar", "Polybar"
WM_NAME(STRING) =
_NET_WM_STRUT_PARTIAL(CARDINAL) = 0, 0, 28, 0, 0, 0, 0, 0, 0, 2559, 0, 0
_NET_WM_STRUT(CARDINAL) = 0, 0, 28, 0
```

Thanks

**Reply by mattmart3** [reply]
So if I understand it correctly the `_NET_WM_DESKTOP(CARDINAL)` property should be `-1` or `0xFFFFFFFF`. The property looks correct for the other windows: it reflects the workspace of where the window currently is and it changes dynamically if I move a window around the workspaces.

Even if I change the value set by polybar that property always reflect the workspace where polybar is launched from. For instance:

```diff
diff --git a/src/components/bar.cpp b/src/components/bar.cpp
index cacf2fb3..4462c082 100644
--- a/src/components/bar.cpp
+++ b/src/components/bar.cpp
@@ -624,7 +624,7 @@ void bar::reconfigure_wm_hints() {
   ewmh_util::set_wm_state(win, {_NET_WM_STATE_STICKY, _NET_WM_STATE_ABOVE});

   m_log.trace("bar: Set window _NET_WM_DESKTOP");
-  ewmh_util::set_wm_desktop(win, 0xFFFFFFFF);
+  ewmh_util::set_wm_desktop(win, 8);

   m_log.trace("bar: Set window _NET_WM_PID");
   ewmh_util::set_wm_pid(win);
```

Could it be that the WM just overrides that value? How should the WM behave instead?

**Reply by patrick96** [reply]
> So if I understand it correctly the _NET_WM_DESKTOP(CARDINAL) property should be -1 or 0xFFFFFFFF.

Exactly. For me, `xprop` shows `_NET_WM_DESKTOP(CARDINAL) = 4294967295`.

> Could it be that the WM just overrides that value? How should the WM behave instead?

Yes, looks like it doesn't propagate this special value properly.

For reference, the spec for this is [here](https://specifications.freedesktop.org/wm-spec/wm-spec-1.3.html#idm46072576096880).
The correct behavior would be that windows that set `0xFFFFFFFF` for `_NET_WM_DESKTOP` should keep that value.

**Reply by mattmart3** [reply]
Great, thanks for the help. I will try to address this to LeftWM.
